File: app/src/main/java/app/simit/com/motivationalquotes/ui/quotes/QuotesFragment.java
package app.simit.com.motivationalquotes.ui.quotes;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import app.simit.com.motivationalquotes.Quote;
import app.simit.com.motivationalquotes.QuoteAdapter;
import app.simit.com.motivationalquotes.R;
import app.simit.com.motivationalquotes.Utils;

public class QuotesFragment extends Fragment {
    private static final String TAG = "QuotesFragment";
    private QuotesViewModel quotesViewModel;
    private RecyclerView recyclerView;
    private QuoteAdapter quoteAdapter;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        quotesViewModel =
                ViewModelProviders.of(this).get(QuotesViewModel.class);
        View root = inflater.inflate(R.layout.fragment_quotes, container, false);
        recyclerView =(RecyclerView) root.findViewById(R.id.allQuotesRecyclerView);
        quoteAdapter = new QuoteAdapter(getActivity());
        ArrayList<Quote> quotes= Utils.getQuotes();
        Log.d(TAG, "onCreateView: Quotes data: " +quotes.toString());
        if(null!=quotes){
            quoteAdapter.setQuotes(quotes);
        }
        recyclerView.setAdapter(quoteAdapter);
        recyclerView.setLayoutManager(new GridLayoutManager(getActivity(),2));

        return root;
    }
}

File: app/src/main/java/app/simit/com/motivationalquotes/ui/quotes/QuotesFragment.kt
package app.simit.com.motivationalquotes.ui.quotes

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import app.simit.com.motivationalquotes.Api.QuoteCalls
import app.simit.com.motivationalquotes.ui.quotes.QuoteList_.QuoteAdapter
import app.simit.com.motivationalquotes.databinding.FragmentQuotesBinding
import app.simit.com.motivationalquotes.ui.quotes.QuoteList_.QuoteListDecoretor
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class QuotesFragment : Fragment() {
    private var decoretor: QuoteListDecoretor? = null
    private lateinit var mViewModel: QuotesViewModel
    private lateinit var quoteAdapter: QuoteAdapter
    private lateinit var binding: FragmentQuotesBinding

    @Inject
    lateinit var quoteCalls: QuoteCalls
    private var RetriveJob: Job? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mViewModel = ViewModelProvider(this).get(QuotesViewModel::class.java)
        mViewModel.setService(quoteCalls)
        Log.i(TAG, "onCreate: ")
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?, savedInstanceState: Bundle?): View? {
        Log.i(TAG, "onCreateView: ")
        binding = FragmentQuotesBinding.inflate(inflater, container, false)

        quoteAdapter = QuoteAdapter(requireContext())

        val layoutManager = StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
        layoutManager.gapStrategy = StaggeredGridLayoutManager.GAP_HANDLING_MOVE_ITEMS_BETWEEN_SPANS
        binding.allQuotesRecyclerView.layoutManager = layoutManager
        binding.allQuotesRecyclerView.adapter = quoteAdapter

        if (decoretor != null) {
            binding.allQuotesRecyclerView.removeItemDecoration(decoretor!!)
        }
        decoretor = QuoteListDecoretor(requireContext())
        binding.allQuotesRecyclerView.addItemDecoration(decoretor!!)


        RetriveJob?.cancel()
        RetriveJob = lifecycleScope.launch {
            mViewModel.getQuotes()?.collectLatest {
                quoteAdapter.submitData(it)
            }
        }


//        mViewModel.quotes.observe(viewLifecycleOwner, Observer { list ->
//            if (list.isNotEmpty()) {
//                quoteAdapter.submitData(list)
//
//                binding.allQuotesRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
//                    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
//                        super.onScrolled(recyclerView, dx, dy)
//                    }
//
//                    override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
//                        mViewModel.setQuotePos(binding.allQuotesRecyclerView.layoutManager?.onSaveInstanceState())
//                        super.onScrollStateChanged(recyclerView, newState)
//                    }
//                })
//
//            } else {
//                // TODO: 6/3/2020 empty state
//            }
//        })

        return binding.root
    }

    companion object {
        private const val TAG = "QuotesFragment"
    }
}